boards: mimxrt700_evk: Extend HiFi4 DSP domain support

VitekST opened the pull request:
This PR includes various bits and pieces (platform setup, driver instantiations, ...) for the `mimxrt700_evk/mimxrt798s/hifi4` domain, so that it can actually do something useful. That target will be supported with samples in the `samples/boards/nxp/adsp/rtxxx/` directory.

Hard prerequisites:

- [x] https://github.com/zephyrproject-rtos/zephyr/pull/93645

Soft prerequisites:

- [x] https://github.com/zephyrproject-rtos/zephyr/pull/93644

VitekST committed:
drivers: nxp,rtxxx-dsp-ctl: Extend to mimxrt798s/hifi4

Rework the driver so that it can cover multiple variants. Add variant DT
bindings. Change the compatible name for the mimxrt685s/cm33 DT.

This needed to be done because the hardware initialisation routines
(power, clocks, ...) are different from mimxrt685s/hifi4 to
mimxrt798s/hifi4. The same is expected for the /hifi1 domain.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
dt: mimxrt798s/cm33_cpu0: Instantiate nxp,rtxxx-dsp-ctrl

Instantiate the nxp,rtxxx-dsp-ctl
driver (nxp,rt700-dsp-ctrl-hifi4 variant) in the mimxrt798s/cm33_cpu0's
DT. Enable it in mimxrt700_evk/mimxrt798s/cm33_cpu0.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
soc: mimxrt798s/hifi4: Rework linker and memory defs

Rename symbols so that they reflect purpose of defined memory regions.
Point region symbols to nodes in the DT. Move bogus IDT section before
DSP's ITCM. Move common ROM and RAM sections before the heap.

The move had to be done as these sections and the heap did overlap.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
boards: mimxrt700_evk: Enable NS & NP GPIO0 access

Enable the GPIO0 peripheral regardless of DT status. Program the
peripheral to allow non-secured and non-privileged accesses.

This is to enable access to the GPIO0 peripheral (both port reads/writes
and interrupts) from the mimxrt798s/hifi4 domain, as that domain is
non-privileged.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
dt: mimxrt798s/cm33_cpu0: Instantiate MU4

Instantiate the MU4 peripheral in SoC's DT. Enable it in board's DT.
Switch variant of the fsl_mu driver included.

The MU4 peripheral is used for IPC with the mimxrt798s/hifi4 domain.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
soc: mimxrt798s/hifi4: Add soc.c

Add soc.c. Handle IRQ mapping (INPUTMUX) and clock setup.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

dbaluta commented on dts/arm/nxp/nxp_rt7xx_cm33_cpu0.dtsi:
```diff
@@ -1201,6 +1201,18 @@
 	};
+
+	mbox_hifi4: mbox@189000 {
+		#mbox-cells = <1>;
```
I think we can call this mbox4 just to be consistent with the rest of the dtses reference for mbox nodes.

dcpleung commented:
One of the commit says DNM so adding the DNM label for now.

VitekST commented:
@dcpleung Thanks. I've eliminated that state of things so the label can be removed.

mmahadevan108 commented on soc/nxp/imxrt/imxrt7xx/hifi4/soc.c:
```diff
@@ -0,0 +1,148 @@
+#endif
+}
+
+void soc_early_init_hook(void)
```
Please ensure the appropriate `Kconfig`  is enabled. Else this will not get invoked

dbaluta commented on boards/nxp/mimxrt700_evk/board.cmake:
```diff
@@ -13,6 +13,8 @@ elseif(CONFIG_SOC_MIMXRT798S_CM33_CPU1)
   board_runner_args(linkserver  "--device=MIMXRT798S:MIMXRT700-EVK")
   board_runner_args(linkserver  "--core=cm33_core1")
+elseif(CONFIG_SOC_MIMXRT798S_HIFI4)
+  board_runner_args(jlink "--device=MIMXRT798S_HiFi4")
```
Does this make sense to be in this patch? It looks unrelated to the rest of the definitions?

VitekST replied:
@dbaluta You're right, it's a board definition, not an SoC definition. Fine, will break that change out.

The J-Link target name suffix is correct - `_HiFi4`. So... if that answers the question, I guess the former.

VitekST committed:
soc: mimxrt798s: Expand definitions for /hifi4

Add desired Kconfig implications for the mimxrt798s/hifi4 domain. Add
pinctrl_soc.h and set up an include path for it.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

VitekST committed:
boards: mimxrt700_evk: Add jlink runner args for hifi4

Add jlink runner arguments for mimxrt700_evk/mimxrt798s/hifi4, so that
domain can be debugged.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>

iuliana-prodan commented on dts/arm/nxp/nxp_rt7xx_cm33_cpu0.dtsi:
```diff
@@ -31,6 +32,34 @@
+			type = <NXP_RTXXX_DSP_REGION_RESET>;
+		};
+
+		/* 1st half of reserved HiFi 4 SRAM region (see below) */
```
Yes, I realized that. I was referring to` HiFi 4` vs `HiFi4`.
Maybe is jut me used to HiFi4, HiFi1, etc.

iuliana-prodan replied:
Above at line 44 , you have it without space, in other places as well, so, I say to have it `HiFi4`.
But this is no blocker for me, just a "nice to have".

iuliana-prodan commented on soc/nxp/imxrt/imxrt7xx/Kconfig:
```diff
@@ -77,4 +82,43 @@ config MCUX_CORE_SUFFIX
 
+if SOC_MIMXRT798S_HIFI4
+
+DT_ADSP_RESET_MEM := $(dt_nodelabel_path,adsp_reset)
```
I noticed that the same setup is present in both RT500 (which I assume is why you added the copyright) and RT600.
It might be helpful to add a comment in the code indicating that these were expected to be defined in the dtsi.

dcpleung replied:
Maybe add a `BUILD_ASSERT()` if it is possible?

VitekST replied:
@dcpleung Assertions. Excellent idea, thanks! I think the proper place for those is `soc/imxrt/imxrt7xx/hifi4/include/soc/memory.h`, but I had troubles adding that macro there as that header file isn't really a C header file but a linker header file. I've added regular conditionals there instead.

iuliana-prodan replied:
I believe it would be better to take the values directly from `dtsi` in `memory.`h - see other examples, like https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/intel/intel_adsp/ace/include/adsp_memory.h#L15
This way you don't need to define so many configs in Kconfig, and used them later in `memory.h`, etc.

dcpleung replied:
If you want to use `BUILD_ASSERT()` to make sure device nodes are defined, you can always create a new C file and put the assertions there.

iuliana-prodan commented on drivers/misc/nxp_rtxxx_dsp_ctrl/nxp_rtxxx_dsp_ctrl.c:
```diff
@@ -78,41 +79,57 @@ static int dsp_ctrl_load_section(const struct device *dev, const void *base, siz
+		POWER_DisablePD(kPDRUNCFG_PD_VDD2_DSP);                                            \
+		POWER_ApplyPD();                                                                   \
+                                                                                                   \
+		CLOCK_SetClkDiv(kCLOCK_DivDspClk, 1U);                                             \
```
Is there a reason you didn’t apply the same approach for RT600 above?

VitekST replied:
@iuliana-prodan No, I was copying code from the MCUXpresso SDK in that case. In this case I had to come up with a configuration.

VitekST committed:
dt: mimxrt700_evk/hifi4: Add definitions

Add dummy interrupt controller, clock control, pin control, Flexcomm 0,
Flexcomm 2, SAI0, SAI1, SAI2 into SoC's DT. Enable relevant nodes in
board's DT and include pinctrl definitions. Add default LED and button
nodes. Set /hifi4's real frequency. Add memory nodes for device's main
SRAM.

Signed-off-by: Vit Stanicek <vit.stanicek@nxp.com>
